### ISSUE TITLE
Writes which failed at replica layer may reappear after restart leading to resurrection of deleted data

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#6082]
Currently, writes are first applied to the commit log and then to the memtable. If the commit log part succeeds but the memtable part fails, then the write will not be visible in the system but may reappear on restart after commit log is replayed.

The memtable part may fail, for example, when the write times out while waiting for admission due to lack of room in the memtable space (blocked on sstable flush). 

This is problematic because:

1) Compaction may drop tombstones relevant for such hidden writes, leading to data resurrection. We already have a similar problem with writes in memtables: https://github.com/scylladb/scylla/issues/4734

1) Repair won't see those writes, which may lead to data resurrection

1) streaming won't see those writes, which makes https://github.com/scylladb/scylla/issues/5251 more likely to run into

nodetool cleanup shouldn't be a problem because it's supposed to flush all memtables and thus discard all commit log segments. It currently doesn't, that's https://github.com/scylladb/scylla/issues/1239

Problems 1) and 2) could be fixed by forcing flush of commit log segments which are old enough, before gc grace period elapses for the writes it may contain.

Problem 3) could be avoided by fixing https://github.com/scylladb/scylla/issues/5251 (which is now put under https://github.com/scylladb/scylla/issues/5252)

Alternatively, we could ensure that failed writes never reappear. This is discussed below.
The commit log part must come before the write is made visible in the memtable. The reason for that is that we want to provide monotonic reads, so we should not publish uncommitted writes which may disappear on failure. Memtable flush would have to wait with discarding commit log segments until all replay positions are settled, but that's not a problem.

Note that replica-level monotonic read property is not needed to provide cluster-level monotonic reads for CL=QUORUM. It's also not needed for linearizability of LWT because that one replays failed commits before reading. Maybe we could lift that.

If we want to keep monotonic reads, things are more complicated. We would have to ensure that after successful application to the commit log the memtable part must succeed or the node aborts. We should make sure that this abort never happens or is very unlikely. The most common failure would be timing out on available room in the memtable space. We would have to move this waiting before the commit log part. We need to copy the mutation to the memtable space immediately when we're admitted in order to make a reservation. The problem with that is that even though we reserve dirty memory in the memtable space, it is not flushable until we publish the write, after the commit log part completes. We cannot really start flushing memtables with incomplete entries. We cannot finish those entries concurrently with flushing, memtable must be immutable at that point.

**Comment by kbr-** [comment]
> replica-level monotonic read property 

Could you more precisely describe what do you mean by that?

I'm thinking about a write that reached some replica, and then was observed by anyone, using e.g. CL=ALL. The write could have been a failed write, it doesn't matter.
The property would then say that any further CL=ALL reads must also observe the write.

I was thinking whether such property can be at all guaranteed in presence of topology changes. And if it can, whether it can be guaranteed without big sacrifices of availability. Consider this scenario:
1. the replicas for partition 0 are {A, B, C}. Bootstrapping is in progress. Pending replicas are {D}. After bootstrapping, replicas for 0 will be {B, C, D}. Streaming is copying data from A to D. All writes are being sent with a higher RF.
2. a write was made with whatever CL and failed. The write reached only A. Streaming won't copy the write, because it's long past key 0.
3. someone made a read and contacted A. He observed the write. But the write will be forever lost.

If we wanted to preserve the write that landed on A for future monotonic reads, we would somehow have to make sure it gets copied to one of the other replicas.
I don't see any other solution than A starting to reject writes at some point, and some process copying any data that resides only on A to one of the other replicas.

**Comment by tgrabiec** [comment]
> > replica-level monotonic read property
> 
> Could you more precisely describe what do you mean by that?

"replica-level monotonic reads" holds if all writes [*] returned by a replica-level reader created at some moment in time are also seen by any replica-level reader created later on the same table on the same replica.

replica-level reader is a reader which combines shard-level readers from all shards on a given replica for a given table. A shard-level reader is a reader created using `table::make_reader()`.

A reader obtained with `make_multishard_streaming_reader()` is a replica-level reader. So are the readers used under the hood by `database::query()`.

[*] That is, of course, ignoring writes subject to the standard garbage-collection rules (old-enough deletions are permitted to disappear).

> 
> I'm thinking about a write that reached some replica, and then was observed by anyone, e.g. a read that used CL=ALL. The write could have been a failed write, it doesn't matter.
> The property would then say that any further CL=ALL reads must also observe the write.
> 
> I was thinking whether such property can be at all guaranteed in presence of topology changes. And if it can, whether it can be guaranteed without big sacrifices of availability. Consider this scenario:
> 
> 1. the replicas for partition 0 are {A, B, C}. Bootstrapping is in progress. Pending replicas are {D}. After bootstrapping, replicas for 0 will be {B, C, D}. Streaming is copying data from A to D. All writes are being sent with a higher RF.
> 2. a write was made with whatever CL and failed. The write reached only A. Streaming won't copy the write, because it's long past key 0.
> 3. someone made a read and contacted A. He observed the write. But the write will be forever lost.
> 
> If we wanted to preserve the write that landed on A for future monotonic reads, we would somehow have to make sure it gets copied to one of the other replicas.
> I don't see any other solution than A starting to reject writes at some point, and some process copying any data that resides only on A to one of the other replicas.

You're talking about cluster-level monotonic reads. This is currently achieved by repairing mismatched replicas before responding with read results. See `abstract_read_executor::reconcile()` where it calls `schedule_repair()`.

We don't need replica-level monotonic reads for cluster-level monotonic reads, but something weaker. We only need replica-level read-your-writes, which means that successful writes are seen by later reads.

**Comment by kbr-** [comment]
> "replica-level monotonic reads" holds if all writes [*] returned by a replica-level reader created at some moment in time are also seen by any replica-level reader created later on the same table on the same replica.
> 
> replica-level reader is a reader which combines shard-level readers from all shards on a given replica for a given table. A shard-level reader is a reader created using table::make_reader().
> 
> A reader obtained with make_multishard_streaming_reader() is a replica-level reader. So are the readers used under the hood by storage_proxy::query_nonsingular_mutations_locally().

Ok, so it sounds like a mostly internal thing. Hence why do we care whether we have replica-level monotonic reads, if that doesn't affect the user (who cares only about cluster-level consistency properties)?
Is is just for the convenience of our internal algorithms, which might want to use such a property?

> You're talking about cluster-level monotonic reads. This is currently achieved by repairing mismatched replicas before responding with read results. See abstract_read_executor::reconcile() where it calls schedule_repair()

Ok, thanks for the clarification.
I'm aware of the read repair mechanism, although I thought it runs only for QUORUM reads (not sure if it works for ALL reads, and for ONE reads it doesn't make much sense).

But read repair won't help us in the scenario I gave, if the read in point 3. is made using ONE consistency level, right? Well, of course ONE reads don't provide read monotonicity, it's pretty much impossible, but the problem I have with this scenario is that some write was observed by someone, but will never again be observed by anyone (even though we had no unrecoverable replica crashes). I'm wondering if that's OK (I'm trying to have a clear picture of what exact consistency properties do we provide to the user).
